File: kotlin/src/test/kotlin/unit/advertisement/domain/model/value_object/AdvertisementDateTest.kt
package unit.advertisement.domain.model.value_object

import advertisement.domain.model.value_object.AdvertisementDate
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.lang.reflect.Modifier
import java.time.LocalDateTime


class AdvertisementDateTest
{
    @Test
    fun testShouldNotBeInstantiatedWithTheConstructor() {
        Assertions.assertThrows(NoSuchMethodException::class.java) {
            Assertions.assertTrue(Modifier.isPrivate(AdvertisementDate::class.java.getDeclaredConstructor().modifiers))
        }
    }

    @Test
    fun testShouldCreateADate() {
        val dateNow = LocalDateTime.now()

        val result = AdvertisementDate.build(dateNow)

        result.fold(
            { error -> Assertions.fail("Expected a valid date, but got error: $error") },
            { id -> Assertions.assertEquals(dateNow, id.value()) }
        )
    }
}

File: php/tests/Unit/Advertisement/Domain/ValueObject/AdvertisementDateTest.php
<?php
declare(strict_types=1);

namespace Tests\Demo\App\Unit\Advertisement\Domain\ValueObject;

use Demo\App\Advertisement\Domain\Model\ValueObject\AdvertisementDate;
use PHPUnit\Framework\TestCase;
use ReflectionClass;

class AdvertisementDateTest extends TestCase
{
    private const string ADVERTISEMENT_CREATION_DATE = '2024-02-03 13:30:23';

    public function testShouldNotBeInstantiatedWithTheConstructor(): void
    {
        $class = new ReflectionClass(AdvertisementDate::class);
        $constructor = $class->getConstructor();

        self::assertTrue($constructor->isPrivate(), 'Constructor is not private');
    }

    public function testShouldBeCreatedWithADate(): void
    {
        $result = AdvertisementDate::build(new \DateTime(self::ADVERTISEMENT_CREATION_DATE));
        $advertisementDate = $result->getOrThrow();

        self::assertEquals(new \DateTime(self::ADVERTISEMENT_CREATION_DATE), $advertisementDate->value());
    }
}
